### PR TITLE
Refactored libraries build process

### PR DESCRIPTION
**Pull request by drichmond**
* Created libraries/libraries.mk, an include fragment for building the Manycore Runtime Libraries.

### CODE REVIEW

**Comment by drichmond** [comment]
This is the PR I'm least certain about.

Benefits: 

- If the .cpp runtime libraries are edited then libbsg_manycore_runtime.so will be rebuilt (previously required `make clean`
- Removed instance of recursive make

Drawbacks: 
- It's more complicated. 
- I don't like that LDFLAGS/CFLAGS/INCLUDES for building libbsg_manycore_runtime are specified in the rule. I wish they could be defined in the body, and then reused. (Not entirely sure how to describe this)

File: libraries/Makefile [hunk 1]
```diff
@@ -21,18 +21,18 @@ include ../environment.mk
...
 
-build: $(LIBRARIES)
+build: $(LIBRARIES_PATH)/libbsg_manycore_runtime.so.1.0
 
-install: build
-	mv libbsg_manycore_runtime.so.1.0 /usr/lib64/
-	ln -sf /usr/lib64/libbsg_manycore_runtime.so.1.0 /usr/lib64/libbsg_manycore_runtime.so.1
-	ln -sf /usr/lib64/libbsg_manycore_runtime.so.1.0 /usr/lib64/libbsg_manycore_runtime.so
+install: $(LIBRARIES_PATH)/libbsg_manycore_runtime.so.1.0
+	mv $(notdir $^) /usr/lib64/
```
Comment (mrutt92) [inline libraries/Makefile]: I recommend using `$<` instead of `$^` here.

File: libraries/libraries.mk [hunk 2]
```diff
@@ -17,43 +17,46 @@ CXXSOURCES += $(LIBRARIES_PATH)/bsg_manycore_responder.cpp
...
+
+$(OBJECTS): INCLUDES  = -I$(LIBRARIES_PATH)
+$(OBJECTS): INCLUDES += -I$(SDK_DIR)/userspace/include
+$(OBJECTS): INCLUDES += -I$(HDK_DIR)/common/software/include
+$(OBJECTS): INCLUDES += -I$(AWS_FPGA_REPO_DIR)/SDAccel/userspace/include
 $(OBJECTS): CFLAGS    = -std=c11 -fPIC -D_GNU_SOURCE $(INCLUDES)
 $(OBJECTS): CXXFLAGS  = -std=c++11 -fPIC -D_GNU_SOURCE $(INCLUDES)
-$(OBJECTS): LDFLAGS   = -lfpga_mgmt
+$(OBJECTS): LDFLAGS   = -lfpga_mgmt -fPIC
+$(OBJECTS): % : $(HEADERS)
```
Comment (mrutt92) [inline libraries/libraries.mk]: not sure why you pattern match here...

Comment (drichmond) [reply libraries/libraries.mk]: Oh, because I experimented with `$(OBJECTS): %.o : %.h`

File: testbenches/simlibs.mk [hunk 3]
```diff
@@ -44,37 +44,33 @@ $(TESTBENCH_PATH)/vcs_simlibs: $(TESTBENCH_PATH)/gen_simlibs.tcl
 
 # Define the COSIM macro so that the DPI Versions of functions are called
 $(OBJECTS): CXXFLAGS += -DCOSIM
+$(OBJECTS): CFLAGS   += -DCOSIM
 
 # libfpga_mgmt will be compiled in $(TESTBENCH_PATH), so direct the linker there
 $(LIBRARIES_PATH)/libbsg_manycore_runtime.so.1.0: LDFLAGS +=-L$(TESTBENCH_PATH) 
 $(LIBRARIES_PATH)/libbsg_manycore_runtime.so.1.0: LDFLAGS +=-Wl,-rpath=$(TESTBENCH_PATH)
-$(LIBRARIES_PATH)/libbsg_manycore_runtime.so.1.0: $(TESTBENCH_PATH)/libfpga_mgmt.so
+$(LIBRARIES_PATH)/libbsg_manycore_runtime.so.1.0: % : $(TESTBENCH_PATH)/libfpga_mgmt.so
```
Comment (mrutt92) [inline testbenches/simlibs.mk]: not sure why you pattern match here...

File: libraries/libraries.mk (hunk 2 above)
Comment (drichmond) [reply libraries/libraries.mk]: Fixed

File: libraries/Makefile (hunk 1 above)
Comment (drichmond) [reply libraries/Makefile]: Fixed

File: testbenches/simlibs.mk (hunk 3 above)
Comment (drichmond) [reply testbenches/simlibs.mk]: Fixed

**Pull request merged**